### PR TITLE
PixelCPEClusterRepair Local Errors Bugfix

### PR DESCRIPTION
**Pull request by OzAmram**
This PR introduces a bug fix to the Cluster Repair CPE to solves the degradation in high pT trackng performance seen recently. The issue was traced back to ClusterRepair not implementing the functionality present in the PixelCPETemplateReco for the errors of the cluster splitting used in the Jet Core cluster splitting. 
This PR also includes a fix for a case in which the ClusterRepair CPE was discrepant with respect to the Template CPE: the errors of edge hits which the 2D reco was not run on which were not being set to the special values used in the template CPE. This fix also defines these errors as constants in the base class rather than having them as two sets of 'magic numbers' in the code. 

The PR is an immediate fix for the high pT tracking issue and the plan is that there will be a longer term restructuring of the Template and ClusterRepair CPE's so as to avoid duplicated code. 

@vberta Has done testing to show that this PR does recover the performance loss in high pT tracking, with slides shown [here](https://indico.cern.ch/event/934809/contributions/4015468/attachments/2101571/3533198/presentation_CRissue_OzFixTest4PR.pdf).

There was also a quick test done on the edge hit errors to ensure that the new PR does indeed match the output of the 1D template reco for edge hits where the 2D reco is not run. Results are [here](https://cernbox.cern.ch/index.php/s/vxbIHE0zUeXBtUz)

@mmusich @tsusa @VinInn @tvami @pmaksim1 @cmantill @mtosi

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLocalTracker/SiPixelRecHits/interface/PixelCPEBase.h [hunk 1]
```diff
@@ -251,6 +251,17 @@ class PixelCPEBase : public PixelClusterParameterEstimator {
   bool DoLorentz_;
   bool LoadTemplatesFromDB_;
 
+  //errors for template reco for edge hits, based on observed residuals from
+  //studies likely done in 2011...
+  const float xEdgeXError_ = 23.0f;
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelRecHits/interface/PixelCPEBase.h:256]: static constexpr ?

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc [hunk 2]
```diff
@@ -631,15 +627,27 @@ LocalError PixelCPEClusterRepair::localError(DetParam const& theDetParam, Cluste
   //--- (never used, in fact: let comment it out, shut up the complains of the static analyzer, and save a few CPU cycles)
   float xerr = 0.0f, yerr = 0.0f;
 
-  //--- Check status of both template calls.
-  if
-    UNLIKELY((theClusterParam.ierr != 0) || (theClusterParam.ierr2 != 0)) {
+  // Check if the errors were already set at the clusters splitting level
+  if (theClusterParam.theCluster->getSplitClusterErrorX() > 0.0f &&
+      theClusterParam.theCluster->getSplitClusterErrorX() < 7777.7f &&
```
Comment (jpata) [inline RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:632]: What is 7777.7f? If we can, let's try to avoid hardcoded magic numbers. Could `std::numeric_limits::max` work, or otherwise making it a static constexpr in the header?

Comment (mmusich) [reply RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:632]: For the record, that number is like that since the dawn of time: https://github.com/cms-cvs-history/RecoLocalTracker-SiPixelRecHits/blob/master/src/PixelCPETemplateReco.cc#L483.

and since AFAIK it's set here:

https://github.com/cms-sw/cmssw/blob/971b61e8ee9899c32c57e629a865b780428a42ec/RecoLocalTracker/SubCollectionProducers/src/JetCoreClusterSplitter.cc#L167-L168

it's not very clear to me from where the `7777.7f` is coming from.  Perhaps for this hotfix, to avoid other unpleasant surprises I would make it `static contexpr` as well and defer to a later PR to find a proper limit.

File: RecoLocalTracker/SiPixelRecHits/src/PixelCPETemplateReco.cc [hunk 3]
```diff
@@ -254,20 +254,82 @@ LocalPoint PixelCPETemplateReco::localPosition(DetParam const& theDetParam, Clus
...
+      edm::LogError("PixelCPETemplateReco") << "@SUB = PixelCPETemplateReco::localPosition"
+                                            << "Should never be here. PixelCPETemplateReco should always be called "
+                                               "with track angles. This is a bad error !!! ";
+
+      theClusterParam.templXrec_ = theDetParam.theTopol->localX(theClusterParam.theCluster->x()) -
+                                   lorentz_drift * micronsToCm;  // rough Lorentz drift correction
+      theClusterParam.templYrec_ = theDetParam.theTopol->localY(theClusterParam.theCluster->y());
+    }
+  } else if UNLIKELY (UseClusterSplitter_ && theClusterParam.templQbin_ == 0) {
+    cout << " PixelCPETemplateReco : We should never be here !!!!!!!!!!!!!!!!!!!!!!" << endl;
```
Comment (jpata) [inline RecoLocalTracker/SiPixelRecHits/src/PixelCPETemplateReco.cc:286]: please remove cout, this could be replaced with edm::LogError or abort the program with exception or assert, as is more suitable.

**Comment by jpata** [comment]
Thanks for the code updates @OzAmram! Also, thanks to @slava77 for suggesting a test on 10859.0 (QCD_Pt_3000_3500_13TeV with 100 events).

Seems like we observe some differences in jet b-tagger outputs, for example the one below. If tracking is modified/improved, I think this can be expected and a retraining will be necessary.

jetsc_recoJetedmRefToBaseProdTofloatsAssociationVector_pfDeepCSVJetTags_probb_RECO_obj_data_12

We also observe differences in jet energy fractions, particularly, a reduction in the amount of jets with HF fractions ~ 0. It's not immediately clear to me if this is what is expected from the details of the Cluster Repair algo post-bugfix. @OzAmram @vberta  can you comment?

jetsc_recoPFJets_ak4PFJetsCHS__RECO_obj_HFEMEnergyFraction

**Comment by slava77** [comment]
> We also observe differences in jet energy fractions, particularly, a reduction in the amount of jets with HF fractions ~ 0. It's not immediately clear to me if this is what is expected from the details of the Cluster Repair algo post-bugfix.

it would be good to remake this plot with |eta|>2.2 or so, to be in the relevant region in eta.
Or, just look at the jets vs eta.
This sample should have (almost) no hard jets in this region. It could be just a case of small more central variation taking the more forward jets below the reco threshold.

**Comment by jpata** [comment]
Ah, you're right, looks more like a downward fluctuation in the number of jets...

jetsc_recoPFJets_ak4PFJets__RECO_obj_eta

jetsc_recoPFJets_ak4PFJets__RECO_obj_HFEMEnergyFraction

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by OzAmram** [comment]
> Seems like we observe some differences in jet b-tagger outputs, for example the one below. If tracking is modified/improved, I think this can be expected and a retraining will be necessary.

Yes I think these sort of changes are expected. We are changing the uncertainties on the pixel hit positions so it likely changes the uncertainty/significance of displaced vertices in addition to the overall tracking changes.

**Comment by jpata** [comment]
+reconstruction

- bugfix for the Cluster Repair high-pt tracking issue
- a large number of minor changes to the reco outputs as expected
- more significant changes to b-tagger outputs are observed, this is expected if pixel hit position uncertainties are modified
- a longer-term restructuring of the code will be done in a follow-up to reduce the amount of duplication

**Comment by OzAmram** [comment]
Given that this a significant bugfix, is the plan to backport it to older releases? @mmusich

**Comment by mmusich** [comment]
@OzAmram, I am not sure a backport is desired? as it will change physics output in closed release cycles. Let's see where the discussion goes at the next PPD general meeting.

**Comment by mtosi** [comment]
I doubt....it would mess up all the UL datasets.
but as pointed by Marco, let see the outcome of the discussion at the PPD

**Comment by qliphy** [comment]
+1

**Comment by jpata** [comment]
Just to note for completeness here, Slava pointed out that we should check DQM plots as well. For WF 10859.0, we observe some changes in MET resolution wrt. genMet, e.g. here for pfMET

[DQMData__Run1__JetMET__Runsummary__METValidation__pfMet__METResolution_GenMETTrue_InMETBins.pdf](https://github.com/cms-sw/cmssw/files/5268750/DQMData__Run1__JetMET__Runsummary__METValidation__pfMet__METResolution_GenMETTrue_InMETBins.pdf)

On closer investigation, there is an event where pfMET goes from 4637 GeV -> 5121 GeV (genMet is 5496 GeV), other changes seem minor (carefully making sure we compare the same events).

I wanted to document this here, I see no reason personally that this would be an issue for this bugfix PR, but let me know if you think otherwise.

**Comment by lathomas** [comment]
Hello, 
What sample does this workflow correspond to?

**Comment by jpata** [comment]
@lathomas 

10859.0_QCD_Pt_3000_3500_13+2018+QCD_Pt_3000_3500_13TeV_TuneCUETP8M1_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano

**Comment by lathomas** [comment]
Ok... I don't see how genmet can be 5.5 TeV there

**Comment by jpata** [comment]
More DQM added here: [dqm.pdf](https://github.com/cms-sw/cmssw/files/5268946/dqm.pdf)

I used `pfMet.sumEt()` for this quick check on one event, the plots are from DQM and probably more informative.

**Comment by lathomas** [comment]
So this is not MET but scalar sum Et, which therefore makes sense. I think a change from 4637 GeV to 5121 GeV related to this PR is fully possible (one very high pt track is being recovered?) so I'd consider this as change as expected.
